### PR TITLE
Implement subscript/superscript text

### PR DESCRIPTION
**Pull request by Bebo-Maker**
i have gone ahead and implemented subscript and superscript text. This is based on the implementation from [this repository](https://github.com/toptensoftware/RichTextKit), which also uses skia to render all types of text.

Should fix #104 

You mentioned in a discussion that you want to rework the way text is rendered, so i dont know if this is the right time/branch to submit this. Hopefully this pull request does at least give an idea on how to implement such feature.

### CODE REVIEW

**Comment by MarcinZiabek** [comment]
Thank you for preparing this pull request 😁 I think that the general implementation looks good. 

As far as I remember, SkiaSharp does not offer font-specific information about exact offset/scale of subscript and superscript. Therefore, we need to assume best default values (as you have done). 

We can consider giving the developer the ability to customize offset/scale. This makes the `.SubScript()` and `.SuperScript()` invocations just helper methods with predefined arguments. I am not sure if we actually need that level of customizability.

However, I am not sure how offset/scale alters general font measurement (baseline, height, top and bottom heights). It is worth checking in Word, how line height behaves when text has applied super/sub script effect.

**Comment by Bebo-Maker** [comment]
> We can consider giving the developer the ability to customize offset/scale. This makes the .SubScript() and .SuperScript() invocations just helper methods with predefined arguments. I am not sure if we actually need that level of customizability.

Should not be a problem we to implement it. Personally i dont think its necessary, but its up to you 😊.
<br>

> However, I am not sure how offset/scale alters general font measurement (baseline, height, top and bottom heights). It is worth checking in Word, how line height behaves when text has applied super/sub script effect.

This is the result in word, which is exactly like QuestPDF behaves as far as i can tell.
Sample
<br>

>  This makes the `.SubScript()` and `.SuperScript()` invocations...

I named the methods `Subscript()` and `Superscript()` but can rename them if you want.

**Comment by MarcinZiabek** [comment]
> I named the methods Subscript() and Superscript() but can rename them if you want.

Microsoft Word uses the same naming as you do. Let's follow the industry standard 😁 

> Edit: Nevermind. This is the result i get with QuestPDF. There is something wrong with superscript, i will investigate it later today.

Measuring fonts is just difficult 😥 Are you ok if this feature waits till next release? This gives us time to figure it out. I am usually trying to not introduce not stuff just before release.

**Comment by Bebo-Maker** [comment]
Sure, we already have the previewer for the april release, i think thats enough for one update 😊. I will keep working on this and get it ready for the next release.

**Comment by MarcinZiabek** [comment]
Be aware that you are working on the branch where I have been additionally testing text shaping. It is possible that you are fighting with bugs that I have introduced somewhere along the way :) It was a very early stage where I was learning how to perform shaping with an acceptable performance.

**Comment by Bebo-Maker** [comment]
I did some further investigation.
I can get the expected result now:
image

Using this code works fine:
```csharp
text.Span("Test", TextStyle.Default.Superscript().BackgroundColor(Colors.Red.Medium));
text.EmptyLine();
text.Line("Test", TextStyle.Default.Superscript().BackgroundColor(Colors.Blue.Medium));
```
but this doesnt work (causes the rendering error i have mentioned above)
```csharp
text.Line("Test", TextStyle.Default.Superscript().BackgroundColor(Colors.Red.Medium));
text.Line("Test", TextStyle.Default.Superscript().BackgroundColor(Colors.Blue.Medium));
```

I am not sure whats causing this, i will backport this pull request to the main branch, so we can be sure the text-shaping branch is not the issue.

**Comment by Bebo-Maker** [comment]
I checked this on the main branch and the issue still persists. I found some other issues while testing which all come down to the same problem. I will try to do my best explaining this below:

We need to measure with a smaller text size, because the width of the text will be smaller when using subscript/superscript. However we still want to use the line height of the text if it would be normal. You can see why we need this in the screenshot i posted above which shows the result in ms word.

However im not sure how to go about this. We need to measure with a smaller font width but the normal font height. Do you have any idea on how to solve this issue?

**Comment by Bebo-Maker** [comment]
Fixed the problem i mentioned above. If this is looking good to you i can backport this pull request to the main branch.

**Comment by MarcinZiabek** [comment]
@Bebo-Maker  Hi 😁 I put several comments in the code review. It is mostly my incompetence in the typography domain, which leads to doubts and many questions. I want to understand and learn as much as I can. I have tried to do research and compare my understanding to your implementation. Please do NOT treat my comments as criticism. I am open to discussion. As always, thank you for your work!

**Comment by Bebo-Maker** [comment]
Hi, i am pretty new to github PRs but im afraid that i cannot see/find your comments? 😅

File: QuestPDF/Infrastructure/TextStyle.cs [hunk 1]
```diff
@@ -43,8 +47,8 @@ internal void ApplyGlobalStyle(TextStyle globalStyle)
             HasGlobalStyleApplied = true;
 
             ApplyParentStyle(globalStyle);
-            PaintKey ??= (FontType, Size, FontWeight, IsItalic, Color);
-            FontMetricsKey ??= (FontType, Size, FontWeight, IsItalic);
+            PaintKey ??= (FontType, Size, FontWeight, FontVariant, IsItalic, Color);
+            FontMetricsKey ??= (FontType, Size, FontWeight, FontVariant, IsItalic);
```
Comment (Bebo-Maker) [inline QuestPDF/Infrastructure/TextStyle.cs:51]: Test

**Comment by Bebo-Maker** [comment]
Thats what im doing. There are no comments displayed on the file tab for me 😅
I only see my test comment.
image

**Comment by MarcinZiabek** [comment]
> Thats what im doing. There are no comments displayed on the file tab for me 😅 I only see my test comment.

My bad! 😥 I wasn't aware that I need to submit the review to make it visible to you. I am not working with GitHub at work so this process here is slightly to me. I hope it works for you now

File: QuestPDF/Fluent/TextStyleExtensions.cs [hunk 2]
```diff
@@ -109,7 +109,29 @@ public static TextStyle ExtraBlack(this TextStyle style)
         {
             return style.Weight(FontWeight.ExtraBlack);
         }
-        
+
+        #endregion
+
+        #region Variant
+        public static TextStyle Variant(this TextStyle style, FontVariant variant)
```
[the remark this answers is not in the document]
Comment (Bebo-Maker) [reply QuestPDF/Fluent/TextStyleExtensions.cs:116]: Should be fixed in #179.

File: QuestPDF/Elements/Text/Items/TextBlockSpan.cs [hunk 3]
```diff
@@ -74,28 +74,40 @@ public virtual void Draw(TextDrawingRequest request)
...
             var fontMetrics = Style.ToFontMetrics();
 
+            var glyphOffsetY = GetVerticalGlyphOffsetForStyle(Style);
+
             Canvas.DrawRectangle(new Position(0, request.TotalAscent), new Size(request.TextSize.Width, request.TextSize.Height), Style.BackgroundColor);
 
             if (!string.IsNullOrWhiteSpace(Text))
             {
                 if (RequiresShaping)
-                    Canvas.DrawShapedText(Text, Position.Zero, Style);
```
[the remark this answers is not in the document]
Comment (Bebo-Maker) [reply QuestPDF/Elements/Text/Items/TextBlockSpan.cs:82]: Should be fixed in #179.

File: QuestPDF/Infrastructure/FontVariant.cs [hunk 4]
```diff
@@ -0,0 +1,9 @@
+﻿namespace QuestPDF.Infrastructure
+{
+  public enum FontVariant
```
[the remark this answers is not in the document]
Comment (Bebo-Maker) [reply QuestPDF/Infrastructure/FontVariant.cs:3]: Should be fixed in #179. I named it FontPosition for now.

File: QuestPDF/Infrastructure/TextStyle.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (Bebo-Maker) [reply QuestPDF/Infrastructure/TextStyle.cs:51]: Should be fixed in #179.

**Comment by Bebo-Maker** [comment]
I have ported this pull request to main (see #179) and did some cleanup which you mentioned in youre code review. I tried my best to port over your concerns which are not adressed yet.

File: QuestPDF/Drawing/FontManager.cs [hunk 5]
```diff
@@ -97,6 +103,14 @@ static SKTypeface GetTypeface(TextStyle style)
             }
         }
 
+        private static float GetTextScale(TextStyle style)
+        {
+            if (style.FontVariant == FontVariant.Superscript || style.FontVariant == FontVariant.Subscript)
+                return 0.65f;
```
[the remark this answers is not in the document]
Comment (Bebo-Maker) [reply QuestPDF/Drawing/FontManager.cs:109]: See [here](https://github.com/QuestPDF/QuestPDF/pull/179/files#r848665216).

File: QuestPDF/Elements/Text/Items/TextBlockSpan.cs [hunk 6]
```diff
@@ -74,28 +74,40 @@ public virtual void Draw(TextDrawingRequest request)
...
             void DrawLine(float offset, float thickness)
             {
                 Canvas.DrawRectangle(new Position(0, offset - thickness / 2f), new Size(request.TextSize.Width, thickness), Style.Color);
             }
         }
+
+        private static float GetVerticalGlyphOffsetForStyle(TextStyle style)
+        {
+            if (style.FontVariant == FontVariant.Superscript)
+                return (style.Size ?? 12f) * -0.35f;
```
[the remark this answers is not in the document]
Comment (Bebo-Maker) [reply QuestPDF/Elements/Text/Items/TextBlockSpan.cs:106]: See [here](https://github.com/QuestPDF/QuestPDF/pull/179/files#r848666219).

**Pull request closed without merging**